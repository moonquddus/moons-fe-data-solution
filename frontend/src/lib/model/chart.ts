/**
 * One thing I would change if I had the time:
 * I'm currently using the BE model for chart data & threads, without transforming it at all
 * Being so tightly-coupled leaves us to the whims of BE, which imo isn't great for stability
 * In an ideal world, FE is just a consumer which could handle any BE service, and vice versa
 */
export type ChartDataReponse = ChartData[]

export type ChartData = Record<ChartDataFeature, number> & {
  country: ChartCountry
}

export type ChartCountry = 'FR' | 'GB' | 'BE' | 'DE' | 'ES' | 'IT'

export const validChartDataFeatures = ['hotdog', 'burger', 'sandwich', 'kebab', 'fries', 'donut'] as const
export type ChartDataFeature = typeof validChartDataFeatures[number]

export interface ChartDataPoint {
  feature: ChartDataFeature
  country: ChartCountry
}

export function assertIsChartDataFeature(feature: string): asserts feature is ChartDataFeature {
  if (['hotdog', 'burger', 'sandwich', 'kebab', 'fries', 'donut'].includes(feature)) return
  console.error('Type issue, feature not valid')
}