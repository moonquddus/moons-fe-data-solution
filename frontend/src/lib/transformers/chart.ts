import { ChartData as ChartJSData, ChartDataset, InteractionItem } from "chart.js"
import { matchThread } from "../helpers/comments"
import { assertIsChartDataFeature, ChartData, validChartDataFeatures } from "../model/chart"
import { ActiveThreadPayload, CommentThread } from "../model/comment"

interface ChartJSTransformReducer {
  labels: string[]
  datasets: ChartDataset<'bar'>[]
}

export function transformDataToChartJSData(data: ChartData[], threads: CommentThread[]): ChartJSData<'bar'> {
  // Apologies, I hate overly complicated reducers but I went past the point of no return haha
  const transformedData: ChartJSTransformReducer = data.reduce(
    (accumulator, singleData) => {
    accumulator.labels.push(singleData.country)
    validChartDataFeatures.forEach((featureName, index) => {
      const thread = matchThread(threads, singleData.country, featureName)
      // This is not the prettiest way to handle this...
      // But I have used the second value to show comments
      // The second value actually shifts the bar, hence I divided it by a large number
      accumulator.datasets[index].data.push([singleData[featureName], (thread?.commentsCount ?? 0) / 10000])
    })
    return accumulator

  }, {
    // this part sets the data structure
    labels: [],
    datasets: validChartDataFeatures.map((featureName, index) => {
      return {
        label: featureName,
        backgroundColor: `hsl(${56 * index} 80% 60%)`,
        data: [],
        datalabels: {
          anchor: 'end',
          color: 'black',
          formatter: (value) => {
            return value[1] > 0 ? `${Math.round(value[1] * 10000)}üó®Ô∏è` : ''
          },
        },
      }
    })
  } as ChartJSTransformReducer) // the dreaded type-casting

  return transformedData
}

export function transformElementToThread(element: InteractionItem, data: ChartData[], threads: CommentThread[]): ActiveThreadPayload {
  const dataset = data[element.index]

  const featureKey = Object.keys(dataset)[element.datasetIndex + 1]
  assertIsChartDataFeature(featureKey)


  return {
    id: matchThread(threads, dataset.country, featureKey)?.id ?? undefined,
    chartDataPoint: {
      country: dataset.country,
      feature: featureKey
    }
  }
}
