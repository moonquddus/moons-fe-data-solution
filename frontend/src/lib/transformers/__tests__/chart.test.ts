import { transformDataToChartJSData } from "../chart"
import { createChartData, createCommentThreads } from "../../testing/fixtures"

describe('transformDataToChartJSData', () => {
  it('should transform api data to a Chart JS model', () => {
    const data = createChartData()
    const threads = createCommentThreads()
    expect(transformDataToChartJSData(data, threads)).toMatchSnapshot()
  })
})
