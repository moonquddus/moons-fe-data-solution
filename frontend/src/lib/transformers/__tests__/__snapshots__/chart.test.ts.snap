// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`transformDataToChartJSData should transform api data to a Chart JS model 1`] = `
Object {
  "datasets": Array [
    Object {
      "backgroundColor": "hsl(0 80% 60%)",
      "data": Array [
        Array [
          24,
          0,
        ],
        Array [
          24,
          0,
        ],
      ],
      "datalabels": Object {
        "anchor": "end",
        "color": "black",
        "formatter": [Function],
      },
      "label": "hotdog",
    },
    Object {
      "backgroundColor": "hsl(56 80% 60%)",
      "data": Array [
        Array [
          2642,
          0,
        ],
        Array [
          2642,
          0,
        ],
      ],
      "datalabels": Object {
        "anchor": "end",
        "color": "black",
        "formatter": [Function],
      },
      "label": "burger",
    },
    Object {
      "backgroundColor": "hsl(112 80% 60%)",
      "data": Array [
        Array [
          42,
          0,
        ],
        Array [
          42,
          0,
        ],
      ],
      "datalabels": Object {
        "anchor": "end",
        "color": "black",
        "formatter": [Function],
      },
      "label": "sandwich",
    },
    Object {
      "backgroundColor": "hsl(168 80% 60%)",
      "data": Array [
        Array [
          23,
          0.0035,
        ],
        Array [
          23,
          0,
        ],
      ],
      "datalabels": Object {
        "anchor": "end",
        "color": "black",
        "formatter": [Function],
      },
      "label": "kebab",
    },
    Object {
      "backgroundColor": "hsl(224 80% 60%)",
      "data": Array [
        Array [
          246,
          0,
        ],
        Array [
          246,
          0,
        ],
      ],
      "datalabels": Object {
        "anchor": "end",
        "color": "black",
        "formatter": [Function],
      },
      "label": "fries",
    },
    Object {
      "backgroundColor": "hsl(280 80% 60%)",
      "data": Array [
        Array [
          1,
          0,
        ],
        Array [
          1,
          0.0002,
        ],
      ],
      "datalabels": Object {
        "anchor": "end",
        "color": "black",
        "formatter": [Function],
      },
      "label": "donut",
    },
  ],
  "labels": Array [
    "GB",
    "IT",
  ],
}
`;
